google/cloud/internal/async_rpc_details.h

```diff
@@ -73,7 +74,8 @@ class AsyncUnaryRpcFuture : public AsyncGrpcOperation {
     if (!ok) {
       // `Finish()` always returns `true` for unary RPCs, so the only time we
       // get `!ok` is after `Shutdown()` was called; treat that as "cancelled".
-      promise_.set_value(Status(StatusCode::kCancelled, "call cancelled"));
+      promise_.set_value(internal::CancelledError("gRPC async unary RPC failed",
+                                                  GCP_ERROR_INFO()));
```

Comment:
I have not used this, but we would want something like:

`GCP_ERROR_INFO().WithMetadata("gl-cpp.error.origin", "client")`

https://github.com/googleapis/google-cloud-cpp/blob/226c607e572d43d35647f2a07f453ca1a00d19b9/google/cloud/internal/rest_stub_helpers.cc#L41

Reply:
Thanks for the pointer. Done.

Added a test too.